netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java

```diff
@@ -334,16 +335,18 @@ protected ClientTransportFactory buildTransportFactory() {
     TransportCreationParamsFilterFactory transportCreationParamsFilterFactory =
         dynamicParamsFactory;
     if (transportCreationParamsFilterFactory == null) {
-      SslContext localSslContext = sslContext;
-      if (negotiationType == NegotiationType.TLS && localSslContext == null) {
-        try {
-          localSslContext = GrpcSslContexts.forClient().build();
-        } catch (SSLException ex) {
-          throw new RuntimeException(ex);
+      ProtocolNegotiator negotiator = this.protocolNegotiator;
+      if (negotiator == null) {
+        SslContext localSslContext = sslContext;
+        if (negotiationType == NegotiationType.TLS && localSslContext == null) {
+          try {
+            localSslContext = GrpcSslContexts.forClient().build();
+          } catch (SSLException ex) {
+            throw new RuntimeException(ex);
+          }
         }
+        negotiator = createProtocolNegotiatorByType(negotiationType, localSslContext);
```

Comment:
One other thing;  This is a little wierd.  If no PN is set, each build channel will get a unique PN.  If one has been set, each built channel will get the exact same one.  That's a slightly surprising behavior change.

Reply:
Hmm... true, and that would apply to ALTS as well. Do you think the builder should allow setting a factory for the PN?

Reply:
The thought had occurred; it does make things more complicated since construction would be deferred.   What would you expect?   A new one each time, or it to be reused?

Reply:
So most of the time it looks like this will be per-Channel or could be made per-Channel easily. In order to not complicate things much, when I add the close/shutdown I'll also add a method called when a new Channel has been built. So I guess that becomes a ref/unref sort of style. I expect most implementations to create a new instance per-Channel and just ignore the ref method.